core/shared/src/main/scala-2.13+/ChunkLike.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+import scala.collection.IterableFactoryDefaults
+import scala.collection.SeqFactory
+import scala.collection.immutable.{ IndexedSeq, IndexedSeqOps, StrictOptimizedSeqOps }
+import scala.reflect.ClassTag
+
+/**
+ * `ChunkLike` represents the capability for a `Chunk` to extend Scala's
+ * collection library. Because of changes to Scala's collection library in
+ * 2.13, separate versions of this trait are implemented for 2.11 / 2.12 and
+ * 2.13 / Dotty. This allows code in `Chunk` to be written without concern for
+ * the implementation details of Scala's collection library to the maximum
+ * extent possible.
+ *
+ * Note that `IndexedSeq` is not a referentially transparent interface in that
+ * it exposes methods that are partial (e.g. `apply`), allocate mutable state
+ * (e.g. `iterator`), or are purely side effecting (e.g. `foreach`). `Chunk`
+ * extends `IndexedSeq` to provide interoperability with Scala's collection
+ * library but users should avoid these methods whenever possible.
+ */
+trait ChunkLike[+A]
+    extends IndexedSeq[A]
+    with IndexedSeqOps[A, Chunk, Chunk[A]]
+    with StrictOptimizedSeqOps[A, Chunk, Chunk[A]]
+    with IterableFactoryDefaults[A, Chunk] { self: Chunk[A] =>
+
+  /**
+   * Returns a filtered, mapped subset of the elements of this `Chunk`.
+   */
+  override def collect[B](pf: PartialFunction[A, B]): Chunk[B] =
+    collectChunk(pf)
+
+  /**
+   * Returns the concatenation of mapping every element into a new chunk using
+   * the specified function.
+   */
+  override final def flatMap[B](f: A => IterableOnce[B]): Chunk[B] = {
+    val len                    = self.length
+    var chunks: List[Chunk[B]] = Nil
+
+    var i               = 0
+    var total           = 0
+    var B0: ClassTag[B] = null.asInstanceOf[ClassTag[B]]
+    while (i < len) {
+      val chunk = f(self(i)) match {
+        case chunk: Chunk[B] => chunk
+        case other           => Chunk.fromIterable(other.iterator.to(List))
```

Comment:
`Chunk.fromArray(...toArray)` if possible.

If not possible maybe we should directly make a `Chunk.fromIterator` or something since going "through" List will not be efficient.

Reply:
I think adding that method is a good idea.